feat(animations): Provide a way to lazy load the animations

JeanMeche opened the pull request:
As part of #50399, this work here is a proposition to allow the lazy loading of animations code that can shave off up to 16KB gzipped of the main bundle ! 

The PR introduces a new entry point in the `platform-browser` package : `lazy-animations`. This has been necessary because the eager providers (`BROWSER_ANIMATIONS_PROVIDERS` and `BROWSER_NOOP_ANIMATIONS_PROVIDERS`) were retaining symbols from `@angular/animation/browser` making the code splitting & lazy loading impossible. 

In this entry point, you will find the new `provideLazyLoadingAnimations()` function which will provide a new `RendererFactory2`  : `AsyncAnimationRendererFactory`. 

`AsyncAnimationRendererFactory` will create a new `DynamicDelegationRenderer` which  will delegate to the default renderer by default. 
Aside from the renderer, the new factory will lazily import the `@angular/animations/browser`. Once the package is loaded, the factory will switch the delegate of `DynamicDelegationRenderer` to an `AnimationRenderer`. 

The side effect of this implementation is that the rendering might not include the styles from the animations defined in components until the module is loaded. This should an acceptable drawback. 

Also, I'd like to thank @AndrewKushnir for his help on this PR. 

## PR Type
- [x] Feature

## Does this PR introduce a breaking change?

- [x] Nos

AndrewKushnir commented on packages/platform-browser/lazy-animations/src/async_animation_renderer.ts:
```diff
@@ -0,0 +1,189 @@
+  constructor(
+      private doc: Document, private delegate: RendererFactory2, private zone: NgZone,
+      type: 'animations'|'noop',
+      moduleImpl: Promise<FromAnimationBrowserModule> = import('@angular/animations/browser')) {
```
I'd propose kicking off a dynamic import in the `createRenderer` function. We can also do that **if** there is any animations defined on the `@Component` (i.e. `type.data?.['animation']` is defined). WDYT?

devversion commented on WORKSPACE:
```diff
@@ -139,11 +139,11 @@ yarn_install(
-    srcs = ["node_modules"],
-)
+    manual_build_file_contents = """\\\r
+filegroup(\r
```
is this change intentional?

JeanMeche replied:
It was triggered by the linting if I remember correctly

JeanMeche committed:
feat(animations): Add the possibility of lazy loading animations code.

`provideLazyLoadedAnimations()` returns providers which allow the lazy loading of the animation module.

Lazy loading of the animation code can shave off up to 16KB gzipped of the main bundle.

jessicajaniuk commented on goldens/public-api/platform-browser/lazy-animations/index.md:
```diff
@@ -0,0 +1,14 @@
+import { Provider } from '@angular/core';
+
+// @public
+export function provideLazyLoadedAnimations(type?: 'animations' | 'noop'): Provider[];
```
We've had some conversations about the term "Lazy Loading" internally, and we'd like to move away from it. Please rename this to `provideAnimationsAsync`.

jessicajaniuk commented on packages.bzl:
```diff
@@ -93,6 +93,7 @@ DOCS_ENTRYPOINTS = [
     "platform-browser-dynamic",
     "platform-browser-dynamic/testing",
     "platform-browser/animations",
+    "platform-browser/lazy-animations",
```
Let's change this path to `platform-browser/animations/async` instead. This is what we'd like users to write:
`import {provideAnimationsAsync} from '@angular/platform-browser/animations/async';`.

JeanMeche replied:
I like that idea very much 👍

jessicajaniuk commented on packages/animations/browser/src/error_helpers.ts:
```diff
@@ -259,3 +257,51 @@ export function transitionFailed(name: string, errors: Error[]): Error {
+
+
+/**
+ * The following code is a copy of the same from the Core package, please keep in sync.
```
I'm not thrilled about having to duplicate this code from Core, and my understanding is the only reason we're doing this is due to webpack. Can we add a TODO here to remove this once we remove support for Webpack?

JeanMeche replied:
I'm happy to report that I'm not able to reproduce the initial issue. I've removed the duplicate code.

AndrewKushnir commented on packages/animations/browser/src/createEngine.ts:
```diff
@@ -0,0 +1,21 @@
+import {WebAnimationsDriver} from './render/web_animations/web_animations_driver';
+
+export function createEngine(type: 'animations'|'noop', doc: Document): AnimationEngine {
+  if (type === 'noop') {
+    return new AnimationEngine(doc, new NoopAnimationDriver(), new NoopAnimationStyleNormalizer());
+  }
```
In this scenario the `NoopAnimationDriver` and `NoopAnimationStyleNormalizer` would become non-tree-shakable (as they are eagerly referenced). Can we refactor the code to avoid this?

AndrewKushnir replied:
@JeanMeche as a possible solution, we can mimic the `provideNoopAnimations()` function and call it `provideNoopAnimationsAsync()`, so that we create one of the engines depending on the function used.

JeanMeche replied:
About that,  I thought about your suggestion. I doesn't seem applicable to me.

createEngine(), the drivers and the styleNormalizers are part of the lazy loaded bundle so we cannot use providers.

AndrewKushnir replied:
Could you please elaborate a bit on this? My assumption is that having the `provideNoopAnimationsAsync()` function would allow us to reference the noop-related code there, so we don't have to do it in the `provideAnimationsAsync()` function. As an option, we can have a callback there that we define there:

```
      useFactory: (doc: Document, renderer: DomRendererFactory2, zone: NgZone) => {
        // in `provideAnimationsAsync`:
        const getEngine = (m) => m.getEngine();
        // in `provideNoopAnimationsAsync`:
        const getEngine = (m) => m.getNoopEngine();
        return new AsyncAnimationRendererFactory(doc, renderer, zone, type, getEngine);
      },
```

JeanMeche replied:
Sure, 

`NoopAnimationDriver`/`NoopAnimationStyleNormalizer` are part of the animation module, the module we want to lazy load. By referencing anything of that module outside of it, we will break the code splitting/lazy loading, so we can't define them in the providers.  

Currently `createEngine()` is not imported statically but via the dynamic import of `@angular/animations/browser`. That why it doesn't break the lazy loading. 

I have tried using  the `animationType` parameter to determine which of `createEngine` or `createNoopEngine` to call.  But both builder don't seem able to infer that `animationType` is readonly in `AsyncAnimationRendererFactory` (and it doesn't tree shake the unused one).

AndrewKushnir commented on packages/animations/browser/src/render/animation_renderer.ts:
```diff
@@ -0,0 +1,134 @@
+    // be used by which cached renderer
+    const delegate = this.delegate.createRenderer(hostElement, type);
+    if (!hostElement || !type?.data?.['animation']) {
+      let renderer: BaseAnimationRenderer|undefined = this._rendererCache.get(delegate);
+      if (!renderer) {
+        // Ensure that the renderer is removed from the cache on destroy
+        // since it may contain references to detached DOM nodes.
+        const onRendererDestroy = () => this._rendererCache.delete(delegate);
+        renderer =
+            new BaseAnimationRenderer(EMPTY_NAMESPACE_ID, delegate, this.engine, onRendererDestroy);
+        // only cache this result when the base renderer is used
+        this._rendererCache.set(delegate, renderer);
+      }
```
nit: to improve minification (property names are not minified), we can reference `this._rendererCache` once and use a const after that.
```suggestion
      const cache = this._rendererCache;
      let renderer: BaseAnimationRenderer|undefined = cache.get(delegate);
      if (!renderer) {
        // Ensure that the renderer is removed from the cache on destroy
        // since it may contain references to detached DOM nodes.
        const onRendererDestroy = () => cache.delete(delegate);
        renderer =
            new BaseAnimationRenderer(EMPTY_NAMESPACE_ID, delegate, this.engine, onRendererDestroy);
        // only cache this result when the base renderer is used
        cache.set(delegate, renderer);
      }
```

JeanMeche replied:
Interesting knowledge, TIL 👍

AndrewKushnir commented on packages/animations/browser/src/render/animation_renderer.ts:
```diff
@@ -0,0 +1,134 @@
+      return;
+    }
+
+    if (this._animationCallbacksBuffer.length == 0) {
```
Similar comment to improve minification:
```suggestion
    const animationCallbacksBuffer = this._animationCallbacksBuffer;
    if (animationCallbacksBuffer.length == 0) {
```

AndrewKushnir replied:
note: there might be similar cases as well

AndrewKushnir commented on packages/animations/browser/src/render/renderer.ts:
```diff
@@ -5,137 +5,19 @@
-    return this.engine.whenRenderingDone();
-  }
-}
+type FactoryWithListenerCallback = RendererFactory2&
```
nit: let's include "Animation" here for readability:
```suggestion
type AnimationFactoryWithListenerCallback = RendererFactory2&
```

AndrewKushnir commented on packages/animations/browser/src/render/web_animations/web_animations_driver.ts:
```diff
@@ -14,6 +14,10 @@ import {packageNonAnimatableStyles} from '../special_cased_styles';
 import {WebAnimationsPlayer} from './web_animations_player';
 
+/**
+ * Remember animation/browser can be lazy loaded.
+ * This class should not import from @angular/core as it breaks the tree shaking
```
Quick question: if we accidentally introduce a dependency on `@angular/core`, would we catch it with tests as well (size-tracking or symbol-tracking)?

AndrewKushnir commented on packages/platform-browser/animations/src/animation_builder.ts:
```diff
@@ -6,12 +6,14 @@
+/**
+ * @publicApi
+ */
+@Injectable({providedIn: 'root'})
```
Can we extract this into a separate PR?

JeanMeche replied:
There is definitely some work to do on this. 
Pulling the `BrowserAnimationBuilder` breaks the code splitting because it's in the `platform-browser/animations` (and not in the `async` secondary entry we had to created). 

I'll remove this from the scope of this PR.

AndrewKushnir commented on packages/platform-browser/lazy-animations/src/async_animation_renderer.ts:
```diff
@@ -0,0 +1,193 @@
+import type {ɵAnimationRendererFactory as AnimationRendererFactory, ɵcreateEngine as createEngine, ɵAnimationRenderer as AnimationRenderer} from '@angular/animations/browser';
+import {Injectable, NgZone, Renderer2, RendererFactory2, RendererStyleFlags2, RendererType2} from '@angular/core';
+
+type FromAnimationBrowserModule = {
```
```suggestion
type AnimationBrowserModuleImports = {
```

AndrewKushnir replied:
Could you please also add a JSDoc above it to explain why we need this type?

JeanMeche replied:
I rewrote that type + added a JSDOC 👍

jessicajaniuk commented on packages/platform-browser/BUILD.bazel:
```diff
@@ -48,6 +48,7 @@ ng_package(
     deps = [
         ":platform-browser",
         "//packages/platform-browser/animations",
+        "//packages/platform-browser/lazy-animations",
```
Let's change this path, as well, so that it doesn't use `lazy`. Let's go with `animations-async` for consistency.

AndrewKushnir commented on packages/platform-browser/lazy-animations/src/async_animation_renderer.ts:
```diff
@@ -0,0 +1,193 @@
+  /**
+   * @internal
+   */
+  private loadImpl(
+      type: 'animations'|'noop',
+      moduleImpl: Promise<FromAnimationBrowserModule> = import('@angular/animations/browser')):
+      Promise<AnimationRendererFactory> {
+    return moduleImpl
```
nit: for readability, I'd propose avoid using a dynamic import as a default value and assign the default value for the `moduleImpl` at the very beginning of the file:
```suggestion
  private loadImpl(
      type: 'animations'|'noop',
      moduleImpl?: Promise<FromAnimationBrowserModule>): Promise<AnimationRendererFactory> {
    moduleImpl ??= import('@angular/animations/browser');
    return moduleImpl
```

AndrewKushnir replied:
We also accept the `moduleImpl` as a constructor argument. Are we supposed to use it here if it was not provided into this function as an argument? If no, we can probably rename the argument to avoid possible confusion.

jessicajaniuk commented on packages/platform-browser/lazy-animations/src/async_animation_renderer.ts:
```diff
@@ -0,0 +1,193 @@
+
+@Injectable()
+export class AsyncAnimationRendererFactory implements RendererFactory2 {
+  private _moduleLoaded: Promise<AnimationRendererFactory>|null = null;
```
I know we refer to this as the `Browser Animations Module`, but there's a bit of overloading with the term `Module`. We could consider renaming this to be something else that's less overloaded.

AndrewKushnir commented on packages/platform-browser/lazy-animations/src/async_animation_renderer.ts:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    // We need to prevent the DomRenderer to throw an error because of synthetic properties
+    if ('throwOnSyntheticProps' in renderer) {
```
I'd propose this refactoring to make the code more compatible with Closure Compiler optimizations that rename fields:
```suggestion
    if (typeof renderer.throwOnSyntheticProps === 'boolean') {
```

AndrewKushnir commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -146,6 +146,8 @@ export class DomRendererFactory2 implements RendererFactory2, OnDestroy {
 class DefaultDomRenderer2 implements Renderer2 {
   data: {[key: string]: any} = Object.create(null);
 
+  throwOnSyntheticProps = true;
```
I'd propose adding a JSDoc here to explain what this flag represents and why it's `true`.

AndrewKushnir reviewed (changes requested):
@JeanMeche thanks for making progress on this PR, leaving some comments after the initial review with @jessicajaniuk.

JeanMeche commented:
@AndrewKushnir @jessicajaniuk Thank you both for your feedbacks !

I have a great news, for a reason I could not determine, the import of the core module in not an issue anymore when using the webpack builder !

AndrewKushnir commented on packages/animations/browser/src/render/animation_renderer.ts:
```diff
@@ -0,0 +1,136 @@
+    if (animationCallbacksBuffer.length == 0) {
+      queueMicrotask(() => {
+        this._zone.run(() => {
+          this._animationCallbacksBuffer.forEach(tuple => {
+            const [fn, data] = tuple;
+            fn(data);
+          });
+          this._animationCallbacksBuffer = [];
+        });
+      });
+    }
+
+    this._animationCallbacksBuffer.push([fn, data]);
```
```suggestion
          animationCallbacksBuffer.forEach(tuple => {
            const [fn, data] = tuple;
            fn(data);
          });
          this._animationCallbacksBuffer = [];
        });
      });
    }

    animationCallbacksBuffer.push([fn, data]);
```

JeanMeche replied:
About the `createEngine` function, I thought about your suggestion. I doesn't seem applicable to me. 

`createEngine()`, the drivers and the styleNormalizers are part of the lazy loaded bundle so we cannot use providers.

AndrewKushnir reviewed (commented):
> I have a great news, for a reason I could not determine, the import of the core module in not an issue anymore when using the webpack builder !

This is great 🎉

AndrewKushnir commented on packages/animations/browser/src/createEngine.ts:
```diff
@@ -0,0 +1,22 @@
+/**
```
Could you please rename the file to `create_engine.ts`?

AndrewKushnir reviewed (commented):
@JeanMeche thanks for addressing the feedback!

I'm also thinking that it'd be great to add a new integration testing app to have a way to capture the bundle size in the size tracking golden files (so that we confirm that everything works as expected and there are no regressions in the future). We can copy the `integration/animations` app and update it to use the `provideAnimationsAsync` function there.

JeanMeche committed:
refactor(animations): Add integration test for animations/async.

This integration test aims to cover that we do not break the code splitting of the animation module when we use `provideAnimationsAsync()`.

JeanMeche commented:
@AndrewKushnir  I've added the integration tests and checked, they fail when the code split/lazy loading is broken.

JeanMeche commented on goldens/size-tracking/integration-payloads.json:
```diff
@@ -35,6 +35,13 @@
+  "animations-async": {
+    "uncompressed": {
+      "runtime": 2755,
+      "main": 221187,
```
Both apps are quite different, the async one uses the router, I'll update the "eager" one for us to have a good comparaison !

JeanMeche committed:
test(animations): update the integration test

Let's have the same test app for async and eagerly loaded animations.

AndrewKushnir reviewed (approved):
@JeanMeche thanks for addressing the feedback, the change looks great! 👍

jessicajaniuk reviewed (approved):
LGTM!!!! 🎉

jessicajaniuk reviewed (approved):
The issue was resolved in #51910. Once we have G3 issues resolved, this is good to go.

reviewed-for: fw-animations, fw-core, integration-tests, public-api, size-tracking

devversion reviewed (approved):
LGTM for Bazel

dylhunn reviewed (approved):
reviewed-for: public-api

jessicajaniuk commented:
[TGP](https://fusion2.corp.google.com/presubmit/568390727/OCL:568390727:BASE:569020688:1695866485607:d5ce5688/targets)

alxhub commented:
This PR was merged into the repository by commit 81e7f5be0380f58b0d7e74c7e8190c865316becc.

robertIsaac commented:
was it renamed to `provideAnimationsAsync`?

jessicajaniuk commented:
@robertIsaac yes, it was.

robertIsaac commented:
I tried it, but saw size increases instead of decrease
The component that have animation is deferred, can it be related?
I can share a reproduce repo if you wish to investigate further

jessicajaniuk commented:
@robertIsaac Please file an issue with the repro and we can take a look.

JeanMeche commented:
@robertIsaac Please open an new issue with a repro repo and we'll happily investigate that  😊
